File: CodexTempo/CodexUsageProvider.cs
namespace CodexTempo;

public sealed class CodexUsageProvider : IDisposable
{
    public const string CachedSourceName = "Codex App Server (cached)";

    private readonly CodexAppServerClient _appServer = new();
    private readonly CodexUsageReader _sessionReader = new();
    private UsageSnapshot? _lastOfficial;

    public async Task<UsageSnapshot?> ReadLatestAsync(CancellationToken cancellationToken = default)
    {
        var live = await _appServer.ReadLatestAsync(cancellationToken);
        if (live is not null)
        {
            live = StabilizeOfficial(live, _lastOfficial, DateTimeOffset.Now);
            var todayUsed = live.Week is null
                ? null
                : await _sessionReader.EstimateTodayUsedForAsync(live.Week, cancellationToken);
            if (_lastOfficial is not null &&
                DateOnly.FromDateTime(_lastOfficial.CapturedAt.LocalDateTime) == DateOnly.FromDateTime(DateTime.Now) &&
                SameWindow(live.Week, _lastOfficial.Week))
                todayUsed = Math.Max(todayUsed ?? 0, _lastOfficial.TodayUsedPercent ?? 0);

            _lastOfficial = live with { TodayUsedPercent = todayUsed };
            return _lastOfficial;
        }

        var now = DateTimeOffset.Now;
        var preserved = PreserveAfterFailure(_lastOfficial, now);
        if (preserved is not null) return preserved;

        return await _sessionReader.ReadLatestAsync(cancellationToken);
    }

    private static UsageSnapshot StabilizeOfficial(
        UsageSnapshot current,
        UsageSnapshot? previous,
        DateTimeOffset now)
    {
        if (previous is null) return current;
        return current with
        {
            FiveHour = StabilizeWindow(current.FiveHour, previous.FiveHour, now),
            Week = StabilizeWindow(current.Week, previous.Week, now)
        };
    }

    private static LimitWindow? StabilizeWindow(
        LimitWindow? current,
        LimitWindow? previous,
        DateTimeOffset now)
    {
        if (previous is null) return current;
        if (current is null)
            return previous.ResetsAt > now ? previous : null;
        if (!SameWindow(current, previous)) return current;

        // Used percentage is monotonic inside one quota window. A lower value
        // is a stale replica or fallback artifact, so retain the highest
        // confirmed value until the reset timestamp actually changes.
        return current.UsedPercent < previous.UsedPercent ? previous : current;
    }

    private static bool SameWindow(LimitWindow? left, LimitWindow? right) =>
        left is not null &&
        right is not null &&
        left.WindowMinutes == right.WindowMinutes &&
        Math.Abs((left.ResetsAt - right.ResetsAt).TotalSeconds) <= 90;

    private static bool HasActiveWindow(UsageSnapshot snapshot, DateTimeOffset now) =>
        snapshot.FiveHour?.ResetsAt > now || snapshot.Week?.ResetsAt > now;

    private static UsageSnapshot? PreserveAfterFailure(
        UsageSnapshot? previous,
        DateTimeOffset now) =>
        previous is not null && HasActiveWindow(previous, now)
            ? previous with { SourceFile = CachedSourceName }
            : null;

    public static bool RunSelfTest()
    {
        var now = DateTimeOffset.Parse("2026-08-03T12:00:00+08:00");
        var reset = now.AddDays(2);
        var previous = new UsageSnapshot(
            null,
            new LimitWindow(38, 10080, reset),
            now.AddSeconds(-10),
            CodexAppServerClient.SourceName,
            4);
        var staleReplica = previous with
        {
            Week = new LimitWindow(33, 10080, reset.AddSeconds(1)),
            CapturedAt = now
        };
        var advanced = staleReplica with { Week = new LimitWindow(40, 10080, reset) };
        var resetWindow = staleReplica with
        {
            Week = new LimitWindow(1, 10080, reset.AddDays(7))
        };
        var preserved = PreserveAfterFailure(previous, now);

        return StabilizeOfficial(staleReplica, previous, now).Week?.UsedPercent == 38
               && StabilizeOfficial(advanced, previous, now).Week?.UsedPercent == 40
               && StabilizeOfficial(resetWindow, previous, now).Week?.UsedPercent == 1
               && HasActiveWindow(previous, now)
               && preserved?.SourceFile == CachedSourceName
               && preserved.Week?.RemainingPercent == 62
               && PreserveAfterFailure(previous, reset.AddSeconds(1)) is null;
    }

    public void Dispose()
    {
        _appServer.Dispose();
        _sessionReader.Dispose();
    }
}

File: macOS/Sources/CodexTempoMac/CodexUsageProvider.swift
import Foundation

actor CodexUsageProvider {
    static let cachedSourceName = "Codex App Server (cached)"

    private let appServer: CodexAppServerClient
    private let sessionReader: CodexUsageReader
    private var lastOfficial: UsageSnapshot?

    init(
        appServer: CodexAppServerClient = CodexAppServerClient(),
        sessionReader: CodexUsageReader = CodexUsageReader()
    ) {
        self.appServer = appServer
        self.sessionReader = sessionReader
    }

    func readLatest() async -> UsageSnapshot? {
        if let official = await appServer.readLatest() {
            let now = Date()
            let stabilized = Self.stabilize(official, previous: lastOfficial, now: now)
            var todayUsed = await stabilized.week.asyncFlatMap { await sessionReader.estimateTodayUsed(for: $0) }
            if let previous = lastOfficial,
               Calendar.current.isDate(previous.capturedAt, inSameDayAs: now),
               Self.sameWindow(stabilized.week, previous.week) {
                todayUsed = max(todayUsed ?? 0, previous.todayUsedPercent ?? 0)
            }
            let result = UsageSnapshot(
                fiveHour: stabilized.fiveHour,
                week: stabilized.week,
                capturedAt: stabilized.capturedAt,
                source: stabilized.source,
                todayUsedPercent: todayUsed
            )
            lastOfficial = result
            return result
        }

        if let preserved = Self.preserveAfterFailure(lastOfficial, now: Date()) {
            return preserved
        }
        return await sessionReader.readLatest()
    }

    func shutdown() async {
        await appServer.shutdown()
    }

    static func stabilize(_ current: UsageSnapshot, previous: UsageSnapshot?, now: Date) -> UsageSnapshot {
        guard let previous else { return current }
        return UsageSnapshot(
            fiveHour: stabilizeWindow(current.fiveHour, previous: previous.fiveHour, now: now),
            week: stabilizeWindow(current.week, previous: previous.week, now: now),
            capturedAt: current.capturedAt,
            source: current.source,
            todayUsedPercent: current.todayUsedPercent
        )
    }

    static func preserveAfterFailure(_ previous: UsageSnapshot?, now: Date) -> UsageSnapshot? {
        guard let previous,
              previous.fiveHour?.resetsAt ?? .distantPast > now ||
                previous.week?.resetsAt ?? .distantPast > now else { return nil }
        return UsageSnapshot(
            fiveHour: previous.fiveHour,
            week: previous.week,
            capturedAt: previous.capturedAt,
            source: cachedSourceName,
            todayUsedPercent: previous.todayUsedPercent
        )
    }

    private static func stabilizeWindow(_ current: LimitWindow?, previous: LimitWindow?, now: Date) -> LimitWindow? {
        guard let previous else { return current }
        guard let current else { return previous.resetsAt > now ? previous : nil }
        guard sameWindow(current, previous) else { return current }
        return current.usedPercent < previous.usedPercent ? previous : current
    }

    private static func sameWindow(_ left: LimitWindow?, _ right: LimitWindow?) -> Bool {
        guard let left, let right,
              left.windowMinutes == right.windowMinutes else { return false }
        return abs(left.resetsAt.timeIntervalSince(right.resetsAt)) <= 90
    }
}

private extension Optional {
    func asyncFlatMap<T>(_ transform: (Wrapped) async -> T?) async -> T? {
        guard let value = self else { return nil }
        return await transform(value)
    }
}
